frontend/src/cards/ui/InsightDisplay.tsx

```diff
@@ -33,42 +29,67 @@ const InsightDisplay: React.FC<InsightDisplayProps> = ({
 }) => {
   const [insight, setInsight] = useState<string>('')
   const [isGeneratingInsight, setIsGeneratingInsight] = useState<boolean>(false)
+  const [rateLimitReached, setRateLimitReached] = useState<boolean>(false)
 
   const queryResponse = queryResponses[0]
   const validData = queryResponse.getValidRowsForField(shareConfig.metricId)
   const [knownData] = splitIntoKnownsAndUnknowns(validData, demographicType)
 
+  useEffect(() => {
+    async function checkLimit() {
+      if (SHOW_INSIGHT_GENERATION) {
+        const isLimited = await checkRateLimitStatus()
+        setRateLimitReached(isLimited)
+      }
+    }
+
+    checkLimit()
+  }, [])
+
   const handleGenerateInsight = async () => {
-    if (!SHOW_INSIGHT_GENERATION || !knownData.length || !metricIds.length)
+    if (
+      !SHOW_INSIGHT_GENERATION ||
+      !knownData.length ||
+      !metricIds.length ||
+      rateLimitReached
+    )
       return
 
     setIsGeneratingInsight(true)
     try {
       const newInsight = await generateInsight({ knownData, metricIds })
       setInsight(newInsight)
+    } catch (error) {
+      if (error instanceof Error && error.message.includes('rate limit')) {
```

Comment:
this feels not quite right; since everything is typed can we match an exact error message? any time we do sub-string matching I think it opens us up to false positives and other potential issues